gossiper: do_status_check: allow evicting dead nodes from membership with no host_id

bhalevy opened the pull request:
The short series allows do_status_check to handle down nodes that don't have HOST_ID application state.

Fixes #16936

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6158/)
:x: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6158/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7459/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6158/)
#### Build Details:
- Duration: 1 hr 52 min
- Builder: spider1.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6162/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6162/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6162/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7462/)
#### Failed Tests (3/24107):
- [topology_experimental_raft.test_tablets.dev.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6162/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_dev_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.dev.6)
- [test_tablet_cleanup](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6162/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_tablet_cleanup) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_cleanup)
- [test_add_node_with_large_partition4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6162/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_add_node_with_large_partition4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_add_node_with_large_partition4)
#### Build Details:
- Duration: 1 hr 55 min
- Builder: spider6.cloudius-systems.com

bhalevy commented:
https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7462/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/FullDtest___full_split006___test_add_node_with_large_partition4/ looks like just flakiness.

Looking also into the unit test failures...

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6167/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6167/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7467/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6167/)
#### Build Details:
- Duration: 1 hr 39 min
- Builder: spider8.cloudius-systems.com

bhalevy commented:
@kbr-scylla please review

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -757,11 +757,12 @@ future<> gossiper::do_status_check() {
-             && (!get_token_metadata_ptr()->is_normal_token_owner(host_id))) {
+        if (!is_alive && (now > expire_time)) {
+          const auto host_id = get_host_id(*eps);
+          if (!host_id || !get_token_metadata_ptr()->is_normal_token_owner(host_id)) {
```
I'm worried that https://github.com/scylladb/scylladb/issues/16936 indicates a real bug that we have, and this change would hide it.

The bug would be that an endpoint appears in gossiper state which does not have a Host ID.
My question is how can this happen?

I thought it's an invariant we preserve that every endpoint comes with a Host ID.
Whenever node starts gossiping, it always inserts its host ID.
And also after a full cluster restart, when nodes are gossiping their peers' state restored from disk, they also include their host IDs.

So isn't there some larger problem we're dealing with? That host IDs are disappearing?

bhalevy replied:
We can add an internal error in non-release modes to help diagnose the problem in testing and follow-through, as suggested here, in release mode so not to interfere with production.

bhalevy commented:
In v2 ([4f87401](https://github.com/scylladb/scylladb/commit/4f874014fc1993a411c95fe869d8f65fa35649f8)):
- rebased
- gms: endpoint_state: add get_host_id
- gossiper: do_status_check: allow evicting dead nodes from membership
  - use eps->get_host_id
  - call on_internal_error_noexcept if !host_id to abort in testing and follow through in production

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6759/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6759/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7984/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6759/)
#### Failed Tests (4/25267):
- [test_simple_add_new_node_while_schema_changes_with_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6759/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_add_new_node_while_schema_changes_with_repair) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_schema_changes_with_repair)
- [test_simple_add_new_node_while_schema_changes_with_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6759/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_add_new_node_while_schema_changes_with_repair_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_schema_changes_with_repair)
- [test_cassandra_issue_10250](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6759/testReport/junit/%28root%29/test_concurrent_schema/Tests___Unit_Tests___test_cassandra_issue_10250) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cassandra_issue_10250)
- [topology.test_concurrent_schema.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6759/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_concurrent_schema_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_concurrent_schema.debug.2)
#### Build Details:
- Duration: 2 hr 8 min
- Builder: spider1.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6759/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6759/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7984/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6759/)
> 
> #### Failed Tests (4/25267):
> * [test_simple_add_new_node_while_schema_changes_with_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6759/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_add_new_node_while_schema_changes_with_repair) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_schema_changes_with_repair)
> * [test_simple_add_new_node_while_schema_changes_with_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6759/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_add_new_node_while_schema_changes_with_repair_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_schema_changes_with_repair)

https://github.com/scylladb/scylladb/issues/17028

> * [test_cassandra_issue_10250](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6759/testReport/junit/%28root%29/test_concurrent_schema/Tests___Unit_Tests___test_cassandra_issue_10250) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cassandra_issue_10250)
> * [topology.test_concurrent_schema.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6759/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_concurrent_schema_debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_concurrent_schema.debug.2)
> 
> #### Build Details:
> * Duration: 2 hr 8 min
> * Builder: spider1.cloudius-systems.com

https://github.com/scylladb/scylladb/issues/14612#issuecomment-1632291426 / https://github.com/scylladb/scylladb/issues/14312

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6770/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6770/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7996/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6770/)
#### Build Details:
- Duration: 2 hr 3 min
- Builder: spider6.cloudius-systems.com

kbr-scylla commented:
LGTM, but waiting until `next` is unfreezed.

kbr-scylla commented:
> The short series allows do_status_check to handle down nodes that hove HOST_ID application state.

s/hove/don't have/?

Also, could you please rebase the series so we could rerun CI on latest master with dtest-raft-topology too?

bhalevy commented:
- fixed typo in PR description
- rebased
- gossiper: get_host_id: differentiate between no endpoint_state and no application_state
- gossiper: print the host_id when endpoint state goes UP/DOWN

bhalevy commented:
@kbr-scylla I'm not sure about backport.
Please advise

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7264/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7264/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8423/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/389/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7264/)
#### Build Details:
- Duration: 2 hr 3 min
- Builder: spider4.cloudius-systems.com

kbr-scylla commented:
I see no evidence that backport is necessary.
